### PR TITLE
#38 bulk endpoint for updating status

### PR DESCRIPTION
**Pull request by alesyalis**

### CODE REVIEW

File: Recrutify/Recrutify.Host/Controllers/CandidatesController.cs [hunk 1]
```diff
@@ -97,5 +97,14 @@ public async Task<ActionResult> BulkCreateTestFeedbacksAsync(BulkCreateTestFeedb
             await _candidateService.BulkCreateTestFeedbacksAsync(bulkCreateTestFeedbackDTO);
             return NoContent();
         }
+
+        //[Authorize(Policy = Constants.Policies.FeedbackPolicy)]
```
Comment (elizavetapavliv) [inline Recrutify/Recrutify.Host/Controllers/CandidatesController.cs:101]: create new policy ManagerPolicy and use it here

File: Recrutify/Recrutify.Host/Extensions/ValidatorsRegistration.cs [hunk 2]
```diff
@@ -16,6 +16,7 @@ public static void AddValidators(this IServiceCollection services)
             services.AddSingleton<IValidator<ProjectResult>, ProjectResultValidator>();
             services.AddSingleton<IValidator<UpsertFeedbackDTO>, UpsertFeedbackValidator>();
             services.AddSingleton<IValidator<BulkCreateTestFeedbackDTO>, BulkCreateTestFeedbackValidator>();
+            services.AddSingleton<IValidator<BulkUpdateStatusDTO>, BulkUpdateStatusReasonCandidatsValidator>();
```
Comment (elizavetapavliv) [inline Recrutify/Recrutify.Host/Extensions/ValidatorsRegistration.cs:19]: BulkUpdateStatusReasonValidator

File: Recrutify/Recrutify.Services/Services/Abstract/ICandidateService.cs [hunk 3]
```diff
@@ -25,5 +25,7 @@ public interface ICandidateService
         Task<CandidateDTO> GetCandidateWithProjectAsync(Guid id, Guid projectId);
 
         Task BulkCreateTestFeedbacksAsync(BulkCreateTestFeedbackDTO bulkCreateTestFeedbackDTO);
+
+        Task BulkUpdateStatusByIdsAsync(BulkUpdateStatusDTO bulkUpdateStatusDTO);
```
Comment (elizavetapavliv) [inline Recrutify/Recrutify.Services/Services/Abstract/ICandidateService.cs:29]: BulkUpdateStatusReasonAsync

File: Recrutify/Recrutify.Services/Services/CandidateService.cs [hunk 4]
```diff
@@ -152,5 +152,10 @@ public Task BulkCreateTestFeedbacksAsync(BulkCreateTestFeedbackDTO bulkCreateTes
                     Type = FeedbackType.Test,
                 });
         }
+
+        public async Task BulkUpdateStatusByIdsAsync(BulkUpdateStatusDTO bulkUpdateStatusDTO)
```
Comment (elizavetapavliv) [inline Recrutify/Recrutify.Services/Services/CandidateService.cs:156]: BulkUpdateStatusAsync

File: Recrutify/Recrutify.Services/Services/CandidateService.cs [hunk 5]
```diff
@@ -152,5 +152,10 @@ public Task BulkCreateTestFeedbacksAsync(BulkCreateTestFeedbackDTO bulkCreateTes
...
+        {
+            await _candidateRepository.UpdateStatusByIdsAsync(bulkUpdateStatusDTO.CandidatesIds, bulkUpdateStatusDTO.ProjectId, _mapper.Map<Status>(bulkUpdateStatusDTO.Status), bulkUpdateStatusDTO.Reason);
```
Comment (elizavetapavliv) [inline Recrutify/Recrutify.Services/Services/CandidateService.cs:158]: no async/await - return

File: Recrutify/Recrutify.Services/Validators/BulkUpdateStatusReasonCandidatsValidator.cs [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+using Recrutify.Services.DTOs;
+
+namespace Recrutify.Services.Validators
+{
+    public class BulkUpdateStatusReasonCandidatsValidator : AbstractValidator<BulkUpdateStatusDTO>
+    {
+        public BulkUpdateStatusReasonCandidatsValidator()
+        {
+            RuleFor(x => x.CandidatesIds)
+                .NotNull()
```
Comment (elizavetapavliv) [inline Recrutify/Recrutify.Services/Validators/BulkUpdateStatusReasonCandidatsValidator.cs:12]: return candidatesexising method and mustAsync rule

File: Recrutify/Recrutify.Services/Services/CandidateService.cs (hunk 5 above)
Comment (alesyalis) [reply Recrutify/Recrutify.Services/Services/CandidateService.cs:158]: done

File: Recrutify/Recrutify.Services/Services/Abstract/ICandidateService.cs (hunk 3 above)
Comment (alesyalis) [reply Recrutify/Recrutify.Services/Services/Abstract/ICandidateService.cs:29]: done

File: Recrutify/Recrutify.Services/Services/CandidateService.cs (hunk 4 above)
Comment (alesyalis) [reply Recrutify/Recrutify.Services/Services/CandidateService.cs:156]: done

File: Recrutify/Recrutify.Host/Extensions/ValidatorsRegistration.cs (hunk 2 above)
Comment (alesyalis) [reply Recrutify/Recrutify.Host/Extensions/ValidatorsRegistration.cs:19]: done

File: Recrutify/Recrutify.Host/Controllers/CandidatesController.cs (hunk 1 above)
Comment (alesyalis) [reply Recrutify/Recrutify.Host/Controllers/CandidatesController.cs:101]: done

File: Recrutify/Recrutify.Services/Validators/BulkUpdateStatusReasonCandidatsValidator.cs (hunk 6 above)
Comment (alesyalis) [reply Recrutify/Recrutify.Services/Validators/BulkUpdateStatusReasonCandidatsValidator.cs]: done

File: Recrutify/Recrutify.Host/Constants.cs [hunk 7]
```diff
@@ -8,6 +8,7 @@ public static class Policies
             public const string FeedbackPolicy = nameof(FeedbackPolicy);
             public const string AdminPolicy = nameof(AdminPolicy);
             public const string HighAccessPolicy = nameof(HighAccessPolicy);
+            public const string ManagerPolicy = nameof(ManagerPolicy);
```
Comment (elizavetapavliv) [inline Recrutify/Recrutify.Host/Constants.cs:11]: you haven't registered policy in startup

File: Recrutify/Recrutify.Services/Validators/BulkUpdateStatusReasonValidator.cs [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+                .MaximumLength(128)
+                .When(x => x.Status == StatusDTO.Declined);
+            RuleFor(x => x.Status)
+                .IsInEnum()
+                .WithMessage("Choose from existing statuses")
+                .Must(s => new[] { StatusDTO.Accepted, StatusDTO.Declined, StatusDTO.WaitingList, }.Contains(s))
+                .WithMessage("The selected status is not valid");
+        }
+
+        private async Task<bool> CandidatesAreExistingAsync(BulkUpdateStatusDTO dto, CancellationToken cancellationToken)
```
Comment (elizavetapavliv) [inline Recrutify/Recrutify.Services/Validators/BulkUpdateStatusReasonValidator.cs:40]: remove

Comment (alesyalis) [reply Recrutify/Recrutify.Services/Validators/BulkUpdateStatusReasonValidator.cs]: done

File: Recrutify/Recrutify.Services/Validators/BulkUpdateStatusReasonCandidatsValidator.cs [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+            RuleFor(x => x.Reason)
+                .NotNull()
+                .NotEmpty()
+                .MaximumLength(128)
+                .When(x => x.Status == StatusDTO.Declined);
+            RuleFor(x => x.Status)
+                .IsInEnum()
+                .WithMessage("Choose from existing statuses")
+                .Must(s => new[] { StatusDTO.Accepted, StatusDTO.Declined, StatusDTO.WaitingList, }.Contains(s))
+                .WithMessage("he selected status is not valid");
```
Comment (elizavetapavliv) [inline Recrutify/Recrutify.Services/Validators/BulkUpdateStatusReasonCandidatsValidator.cs:23]: Cannot change to selected status manually

Comment (alesyalis) [reply Recrutify/Recrutify.Services/Validators/BulkUpdateStatusReasonCandidatsValidator.cs:23]: done

File: Recrutify/Recrutify.Host/Constants.cs (hunk 7 above)
Comment (alesyalis) [reply Recrutify/Recrutify.Host/Constants.cs:11]: done

**Review by elizavetapavliv — APPROVED** [review]

**Pull request merged**